File: app/tasks/update_rates.py
import sqlite3
import re
from datetime import datetime, timedelta, timezone
import pytz
from datetimerange import DateTimeRange
import os

HA_DB_URL = os.getenv("HA_DB_URL")


def update_import():
    conn = sqlite3.connect(HA_DB_URL)
    conn.row_factory = sqlite3.Row  # enables using dict in rows
    c = conn.cursor()

    # set rates for consumption
    sql = """
        SELECT * 
        FROM Electricity 
        WHERE 
            rateId IS NULL
    """

    usage = c.execute(sql).fetchall()

    for el in usage:
        try:
            dt = datetime.strptime(el["datetime_start"], "%Y-%m-%d %H:%M:%S")
        except ValueError:
            dt = datetime.strptime(el["datetime_start"], "%Y-%m-%d %H:%M:%S%z")
        utc = pytz.timezone("UTC")
        dt = utc.localize(dt)

        hours = dt.strftime("%H")
        mins = dt.strftime("%M")
        el_time = f"{hours}:{mins}"
        date_str = dt.strftime("%Y-%m-%d")

        print(f"el_time {el_time} - {date_str}")

        sql = f"""
            SELECT
                id
            FROM
                Supplier s
            WHERE
                (
                    DATE(?) BETWEEN DATE(
                        s.supplier_start / 1000,
                        'unixepoch',
                        'localtime'
                    ) AND DATE(s.supplier_end / 1000, 'unixepoch', 'localtime')
                )
                OR (
                    DATE(?) >= DATE(
                        s.supplier_start / 1000,
                        'unixepoch',
                        'localtime'
                    )
                    AND s.supplier_end IS NULL
                )
                AND tariff_type = 'import'
        """
        suppl = c.execute(sql, (date_str, date_str)).fetchall()

        if suppl[0]:
            sql = f"""
                SELECT *
                FROM rates
                WHERE
                    supplierId = {suppl[0]["id"]}
            """
            rates = c.execute(sql).fetchall()

            # Define time zones
            utc = pytz.timezone("UTC")
            bst = pytz.timezone("Europe/London")

            for rate in rates:
                if rate["rate_type"] != "fixed":
                    # Get the actual datetime in UTC
                    actual_dt_utc = datetime.strptime(el["datetime_start"], "%Y-%m-%d %H:%M:%S")
                    actual_dt_utc = utc.localize(actual_dt_utc)

                    # Convert to local time (BST)
                    actual_dt = actual_dt_utc.astimezone(bst)

                    # extract hour and minute of rate window
                    regex = r"^(\d\d):(\d\d)$"
                    start = re.search(regex, rate["start_time"])
                    end = re.search(regex, rate["end_time"])
                    start_hour = int(start.group(1))
                    start_min = int(start.group(2))
                    end_hour = int(end.group(1))
                    end_min = int(end.group(2))

                    # build dt_start and dt_end based on actual_dt (in BST)
                    dt_start = actual_dt.replace(hour=start_hour, minute=start_min, second=0, microsecond=0)
                    dt_end = actual_dt.replace(hour=end_hour, minute=end_min, second=0, microsecond=0)

                    # If end time is earlier than start time, it means the range crosses midnight
                    if dt_end.time() < dt_start.time():
                        # Shift dt_start to previous day and dt_end to the next day
                        dt_start = dt_start - timedelta(days=1)
                        dt_end = dt_end + timedelta(days=1)
                    else:
                        # This might be needed for cases where the end time is exactly at midnight
                        dt_end = dt_end - timedelta(seconds=1)

                    # Debug print statements
                    # print(f"rate: {rate['rate_type']}, start_time: {rate['start_time']}, end_time: {rate['end_time']}")
                    # print(f"actual_dt: {actual_dt}, dt_start: {dt_start}, dt_end: {dt_end}")

                    # Use <= for both ends of the range
                    if dt_start <= actual_dt <= dt_end:
                        print(f"Matching rate: {rate['rate_type']}")  # Debug print statement

                        sql = f"""
                            UPDATE
                                electricity
                            SET
                                rateId = ?
                            WHERE
                                id = ?
                        """
                        c.execute(sql, (rate["id"], el["id"]))
                        conn.commit()
                else:
                    sql = f"""
                        UPDATE
                            electricity
                        SET
                            rateId = ?
                        WHERE 
                            id = ?
                    """
                    c.execute(sql, (rate["id"], el["id"]))
                    conn.commit()
                    pass

    conn.close()


def update_export():
    conn = sqlite3.connect(HA_DB_URL)
    conn.row_factory = sqlite3.Row  # enables using dict in rows
    c = conn.cursor()
    # set rates for production
    sql = """
        SELECT * 
        FROM Electricity 
        WHERE 
            exportRateId IS NULL
            /*AND 
            datetime_start >= DATE(
                strftime('%s', 'now', '-7 days'),
                'unixepoch',
                'localtime'
            )*/
    """

    usage = c.execute(sql).fetchall()

    for el in usage:
        try:
            dt = datetime.strptime(el["datetime_start"], "%Y-%m-%d %H:%M:%S")
        except ValueError:
            dt = datetime.strptime(el["datetime_start"], "%Y-%m-%d %H:%M:%S%z")
        utc = pytz.timezone("UTC")
        dt = utc.localize(dt)

        hours = dt.strftime("%H")
        mins = dt.strftime("%M")
        el_time = f"{hours}:{mins}"
        date_str = dt.strftime("%Y-%m-%d")

        print(f"el_time {el_time} - {date_str}")

        sql = f"""
            SELECT
                id
            FROM
                Supplier s
            WHERE
                (
                    DATE(?) BETWEEN DATE(
                        s.supplier_start / 1000,
                        'unixepoch',
                        'localtime'
                    ) AND DATE(s.supplier_end / 1000, 'unixepoch', 'localtime')
                )
                OR (
                    DATE(?) >= DATE(
                        s.supplier_start / 1000,
                        'unixepoch',
                        'localtime'
                    )
                    AND s.supplier_end IS NULL
                )
                AND tariff_type = 'export'
        """
        try:
            suppl = c.execute(sql, (date_str, date_str)).fetchall()

            if suppl[0]:
                sql = f"""
                    SELECT *
                    FROM rates
                    WHERE
                        supplierId = {suppl[0]["id"]}
                """
                rates = c.execute(sql).fetchall()

                # Define time zones
                utc = pytz.timezone("UTC")
                bst = pytz.timezone("Europe/London")

                for rate in rates:
                    if rate["rate_type"] != "fixed":
                        # Get the actual datetime in UTC
                        actual_dt_utc = datetime.strptime(el["datetime_start"], "%Y-%m-%d %H:%M:%S")
                        actual_dt_utc = utc.localize(actual_dt_utc)

                        # Convert to local time (BST)
                        actual_dt = actual_dt_utc.astimezone(bst)

                        # extract hour and minute of rate window
                        regex = r"^(\d\d):(\d\d)$"
                        start = re.search(regex, rate["start_time"])
                        end = re.search(regex, rate["end_time"])
                        start_hour = int(start.group(1))
                        start_min = int(start.group(2))
                        end_hour = int(end.group(1))
                        end_min = int(end.group(2))

                        # build dt_start and dt_end based on actual_dt (in BST)
                        dt_start = actual_dt.replace(hour=start_hour, minute=start_min, second=0, microsecond=0)
                        dt_end = actual_dt.replace(hour=end_hour, minute=end_min, second=0, microsecond=0)

                        # If end time is earlier than start time, it means the range crosses midnight
                        if dt_end.time() < dt_start.time():
                            # Shift dt_start to previous day and dt_end to the next day
                            dt_start = dt_start - timedelta(days=1)
                            dt_end = dt_end + timedelta(days=1)
                        else:
                            # This might be needed for cases where the end time is exactly at midnight
                            dt_end = dt_end - timedelta(seconds=1)

                        # Debug print statements
                        # print(f"rate: {rate['rate_type']}, start_time: {rate['start_time']}, end_time: {rate['end_time']}")
                        # print(f"actual_dt: {actual_dt}, dt_start: {dt_start}, dt_end: {dt_end}")

                        # Use <= for both ends of the range
                        if dt_start <= actual_dt <= dt_end:
                            print(f"Matching rate: {rate['rate_type']}")  # Debug print statement

                            sql = f"""
                                UPDATE
                                    electricity
                                SET
                                    exportRateId = ?
                                WHERE
                                    id = ?
                            """
                            c.execute(sql, (rate["id"], el["id"]))
                            conn.commit()
                    else:
                        sql = f"""
                            UPDATE
                                electricity
                            SET
                                exportRateId = ?
                            WHERE 
                                id = ?
                        """
                        c.execute(sql, (rate["id"], el["id"]))
                        conn.commit()
                        pass
        except Exception as e:
            print(f"Error: {e}")
            pass
    conn.close()


def refresh_db():
    update_import()
    update_export()


if __name__ == "__main__":
    refresh_db()

File: app/web/tasks/update-rates.js
const { PrismaClient } = require("@prisma/client");
const { DateTime } = require("luxon");
// require('dotenv').config();

const prisma = new PrismaClient();

// find the relevant supplier ID based on the date
async function findsupplier_id(dateLocal, tariffType) {
    try {
        const supplier = await prisma.energySupplier.findFirst({
            where: {
                OR: [
                    {
                        AND: [
                            {
                                supplier_start: {
                                    lte: dateLocal,
                                },
                                supplier_end: {
                                    gte: dateLocal,
                                },
                            },
                        ],
                    },
                    {
                        AND: [
                            {
                                supplier_start: {
                                    lte: dateLocal,
                                },
                                supplier_end: null,
                            },
                        ],
                    },
                ],
                tariff_type: tariffType,
            },
            select: {
                id: true,
            },
        });

        return supplier.id;
    } catch (error) {
        console.error("Error finding supplier:", error);
    }
}

const updateImport = async () => {
    try {
        // Fetch records where rateId is null
        const usageRecords = await prisma.gridEnergy.findMany({
            where: {
                rate_id: null,
            },
        });

        for (const record of usageRecords) {
            const rate_id = await computeRateId(record.datetime_start, "import");

            try {
                const updatedRow = await prisma.gridEnergy.update({
                    where: {
                        id: record.id,
                    },
                    data: {
                        rate_id,
                    },
                });
            } catch (e) {
                console.log(e);
            }
        }
    } catch (error) {
        console.error(`Error updating import rates: ${error}`);
    } finally {
        // await prisma.$disconnect();
    }
};

const computeRateId = async (datetime_start, tariffType) => {
    const local_tz = "Europe/London";
    const startDtUtc = DateTime.fromJSDate(datetime_start, { zone: "utc" });
    const startDtLocal = startDtUtc.setZone(local_tz);

    const supplier_id = await findsupplier_id(startDtLocal, tariffType);
    const supplierRates = await prisma.energyRate.findMany({
        where: {
            supplier: {
                id: supplier_id,
            },
        },
    });
    // console.log(supplierRates);

    for (const rate of supplierRates) {
        if (rate.rate_type === "fixed") {
            return rate.id;
        }

        if (rate.rate_type === "off-peak" || rate.rate_type === "peak") {
            const start_str = String(rate.start_time);
            const end_str = String(rate.end_time);
            const start = /^(\d\d):(\d\d)$/.exec(start_str);
            const end = /^(\d\d):(\d\d)$/.exec(end_str);

            let dtStart = startDtLocal.set({
                hour: start[1], minute: start[2], second: 0, millisecond: 0,
            });
            let dtEnd = startDtLocal.set({
                hour: end[1], minute: end[2], second: 0, millisecond: 0,
            });

            if (dtEnd < dtStart) {
                dtStart = dtStart.minus({ days: 1 });
                dtEnd = dtEnd.plus({ days: 1 });
            } else {
                dtEnd = dtEnd.minus({ seconds: 1 });
            }

            if (startDtLocal >= dtStart && startDtLocal <= dtEnd) {
                console.log(`Matching to rate: ${rate.rate_type}`);
                return rate.id;
            }
        }
    }

    return null;
};

const updateExport = async () => {
    try {
        // Fetch records where export_rate_id for export is be null
        const exportRecords = await prisma.electricity.findMany({
            where: {
                export_rate_id: null,
                kwh_exported: { not: null },
            },
        });

        for (const record of exportRecords) {
            const export_rate_id = await computeRateId(record.datetime_start, "export");
            try {
                const updatedRow = await prisma.gridEnergy.update({
                    where: {
                        id: record.id,
                    },
                    data: {
                        export_rate_id,
                    },
                });
                console.log(`Record updated: ${updatedRow.id}`);
            } catch (e) {
                console.error(`Error updating record ID ${record.id}: ${e}`);
            }
        }
    } catch (error) {
        console.error(`Error updating export rates: ${error}`);
    } finally {
        // await prisma.$disconnect();
    }
};

// updateImport();
// updateExport();

module.exports = {
    updateImport,
    updateExport,
};
